drivers/python/rethinkdb/net_twisted.py

```diff
@@ -0,0 +1,360 @@
+        self.buf_token = None
+
+        self.wait_for_handshake = Deferred()
+
+        self._open = True
+
+    def resetBuffer(self):
+        self.buf = ''
+        self.buf_expected_length = 0
+
+    def connectionMade(self):
+        # Send immediately the handshake.
+        self.transport.write(self.factory.handshake_payload)
+        # Defer a timer which will callback when timed out and errback the
+        # wait_for_handshake. Otherwise, it will be cancelled in
+        # handleHandshake.
+        self._timeout_defer = reactor.callLater(self.factory.timeout,
+                self._handleHandshakeTimeout)
+
+    def connectionLost(self, reason):
+        self._open = False
+
+    def _handleHandshakeTimeout(self):
+        # If we are here, we failed to do the handshake before the timeout.
+        # We close the connection and raise an RqlTimeoutError in the
+        # wait_for_handshake deferred.
+        self.transport.loseConnection()
+        self.wait_for_handshake.errback(RqlTimeoutError())
+
+    def _handleHandshake(self, data):
+        try:
+            self.buf += data
+            if self.buf[-1] == b'\0':
+                message = decodeUTF(self.buf[:-1]).split('\n')[0]
+                if message != 'SUCCESS':
+                    # If there is some problem with the handshake, we errback
+                    # our deferred.
+                    self.wait_for_handshake.errback(RqlDriverError('Server'
+                    'dropped connection with message'
+                    '"{msg}"'.format(msg=message)))
+                else:
+                    # We cancel the scheduled timeout.
+                    self._timeout_defer.cancel()
+                    # We callback our wait_for_handshake.
+                    self.wait_for_handshake.callback(None)
+                    # We're now ready to work with real data.
+                    self.state = DatabaseProtocol.READY
+                    self.resetBuffer()
+        except Exception as e:
+            self.wait_for_handshake.errback(e)
+
+    def _handleResponse(self, data):
+        # 1. Read the header, until we read the length of the awaited payload.
+        if self.buf_expected_length == 0:
+            if data >= 12:
+                token, length = struct.unpack('<qL', data[:12])
+                self.buf_token = token
+                self.buf_expected_length = length
+                data = data[12:]
+            else:
+                self.buf += data
+                # We quit the function, it is impossible to have read the
+                # entire payload at this point.
+                return
+
+        # 2. Buffer the data, until the size of the data match the expected
+        # length provided by the header.
+        self.buf += data
+        if len(self.buf) == self.buf_expected_length:
+            self.factory.response_handler(self.buf_token, self.buf)
+            self.buf_token = None
+            self.resetBuffer()
```

Comment:
I don't think this will work if the driver receives two responses in a single dataReceived callback.  It should save the remainder of the buffer (same goes for _handleHandshake, though it's not likely someone will be sending queries before the handshake response is received).